Домашняя работа 4

khitao opened the pull request:
Добавлен сервис WeatherApiService.

Реализована конфигурация RateLimiter.

Создан WebClient для работы с Weather API.

Реализован WeatherApiController для обработки запросов к /weather_api/current.

Добавлены модели данных для парсинга ответа от API.

Реализована обработка исключений API.

Обновлен openapi.yml.

khitao committed:
Домашняя работа 4

Добавлен сервис WeatherApiService.

Реализована конфигурация RateLimiter.

Создан WebClient для работы с Weather API.

Реализован WeatherApiController для обработки запросов к /weather_api/current.

Добавлены модели данных для парсинга ответа от API.

Реализована обработка исключений API.

Обновлен openapi.yml.

madoed commented on SpringBootApp/build.gradle:
```diff
@@ -22,6 +22,10 @@ repositories {
+    implementation 'com.fasterxml.jackson.core:jackson-databind:2.12.3'
+    implementation 'org.springframework.boot:spring-boot-starter-webflux'
+    implementation 'io.github.resilience4j:resilience4j-spring-boot2:1.7.0'
+    implementation 'io.github.resilience4j:resilience4j-reactor:1.7.0'
```
Зависимости для spring-boot-2, но выше spring-boot-3

Соотвественно не будет работать ratelimiter

madoed commented on SpringBootApp/src/main/java/ru/khodov/springbootapp/config/WeatherApiConfig.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    @Bean
+    public WebClient webClient() {
```
Лучше назвать bean более говоряще, типа weatherApiClient

madoed commented on SpringBootApp/src/main/java/ru/khodov/springbootapp/model/Condition.java:
```diff
@@ -0,0 +1,10 @@
+
+@Data
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class Condition {
```
Можно использовать record вместо связки class + @Date

madoed commented on SpringBootApp/src/main/java/ru/khodov/springbootapp/model/Location.java:
```diff
@@ -0,0 +1,14 @@
+    private String region;
+    private String country;
+    private String tz_id;
+    private String localtime;
```
Можно сразу указывать нужный java.time класс с аннотацией JsonFormat

khitao replied:
@JsonFormat(pattern = "yyyy-MM-dd HH:mm")
        LocalDateTime localtime  

Постоянно вылетают такие ошибки JSON decoding error: Cannot deserialize value of type `java.time.LocalDateTime` from String "2023-10-10 4:38": Failed to deserialize java.time.LocalDateTime: (java.time.format.DateTimeParseException) Text '2023-10-10 4:38' could not be parsed at index 11

khitao replied:
Хотя вот, что в документации написано 
![Документ](https://i.imgur.com/l7UfiBd.png)

madoed replied:
Попробуй добавить этот модуль https://fasterxml.github.io/jackson-modules-java8/javadoc/datetime/2.9/com/fasterxml/jackson/datatype/jsr310/JavaTimeModule.html

madoed commented on SpringBootApp/src/main/java/ru/khodov/springbootapp/service/WeatherApiService.java:
```diff
@@ -0,0 +1,81 @@
+    @Value("${weather.api.key}")
+    private String apiKey;
+
+    WebClient webClient;
```
а почему default модификатор доступа?

madoed commented on SpringBootApp/src/main/java/ru/khodov/springbootapp/service/WeatherApiService.java:
```diff
@@ -0,0 +1,81 @@
+
+    public ResponseEntity<WeatherApi> getCurrentWeather(String location) {
+
+        boolean permission = rateLimiter.acquirePermission();
+
+        if (!permission) {
+            throw new ManyRequestsException("Слишком много запросов");
+        }
```
Можно спрятать за аннотацией `@RateLimiter`

khitao replied:
Что то у меня не получается использовать эту аннотацию. Не понимаю, что не так делаю зависимости обновил.(

khitao replied:
Не получается 😩. `@RateLimiter(name = "weatherApi")` не работает. Он почему-то не видит настройки моего Бина в конфигурации и выдает какие-то свои дефолтные ограничения: 2023-10-10T16:52:16.604+03:00 DEBUG 37325 --- [nio-8080-exec-1] i.g.r.s.r.configure.RateLimiterAspect    : Created or retrieved rate limiter 'weatherApi' with period: 'PT0.0000005S'; limit for period: '50'; timeout: 'PT5S'; method: 'ru.khodov.springbootapp.service.WeatherApiService#getCurrentWeather'. Но если в properties прописываю ограничения он их видит.

madoed commented on SpringBootApp/src/main/java/ru/khodov/springbootapp/service/WeatherApiService.java:
```diff
@@ -0,0 +1,81 @@
+                String errorResponseBody = e.getResponseBodyAsString();
+                HttpStatusCode statusCode = e.getStatusCode();
+
+                ObjectMapper objectMapper = new ObjectMapper();
```
На каждую ошибку будет создаваться свой ObjectMapper - слишком затратно, надо вынести бином в конфигурацию

madoed commented on SpringBootApp/src/main/java/ru/khodov/springbootapp/service/WeatherApiService.java:
```diff
@@ -0,0 +1,81 @@
+
+            return new ResponseEntity<>(weatherResponse, HttpStatus.OK);
+
+        } catch (WebClientResponseException e) {
+
+            try {
+                String errorResponseBody = e.getResponseBodyAsString();
+                HttpStatusCode statusCode = e.getStatusCode();
+
+                ObjectMapper objectMapper = new ObjectMapper();
+                WeatherApiError weatherApiError = objectMapper.readValue(errorResponseBody, WeatherApiError.class);
+
+                int errorCode = weatherApiError.getError().getCode();
+                String errorMessage = weatherApiError.getError().getMessage();
+
+                throw new BadRequestToWeatherApiException("Ошибка: " + statusCode + ", Код ошибки API: " + errorCode + ", Описание ошибки API: " + errorMessage, statusCode);
+
+            } catch (JsonProcessingException jsonException) {
+
+                throw new RuntimeException("Ошибка при парсинге JSON");
+
+            }
+
+
+        }
```
Выглядит нечитабельно, нужно выделить в отдельный метод/класс на выбор

Ещё почитать про ErrorHandling https://www.baeldung.com/spring-webflux-errors

madoed reviewed (changes requested):
Нужно доработать по комментариям

khitao committed:
Исправления в дз4

- Обновил зависимости в build.gradle

- Использовал record вместо связки class + @Date

- Исправил ошибку, связанную с модификатором доступа

- Вынес ObjectMapper в конфигурацию

- Исправил ошибку, связанную с модификатором доступа

- Сделал getCurrentWeather читабельным

madoed commented on SpringBootApp/src/main/java/ru/khodov/springbootapp/controllers/WeatherApiController.java:
```diff
@@ -0,0 +1,22 @@
+    private final WeatherApiService weatherApiClient;
+
+    @GetMapping("{location}")
+    public Mono<ResponseEntity<WeatherApi>> getCurrentWeather(@PathVariable String location) {
```
Зачем Mono и ResponseEntity? 
Можно просто оставить Mono<WeatherApi> или просто WeatherApi, сделав на сервисном уровне .block объекту Mono

madoed reviewed (approved):
Доработай комментарии, задачу зачитываю

khitao committed:
Доработки в ДЗ4

Изменил типы возвращаемых значений в методах сервиса и контроллера.

Добавил JavaTimeModule для корректной десериализации LocalDateTime из JSON.
